keps/sig-cluster-lifecycle/draft-20180130-kubeadm-join-master.md

```diff
@@ -0,0 +1,429 @@
+# kubeadm join --master workflow
+
+## Metadata
+
+```yaml
+---
+kep-number: draft-20180130
+title: kubeadm join --master workflow
+status: accepted
+authors:
+  - "@fabriziopandini"
+owning-sig: sig-cluster-lifecycle
+reviewers:
+  - "@errordeveloper"
+  - "@jamiehannaford"
+approvers:
+  - "@luxas"
+  - "@timothysc"
+  - "@roberthbailey"
+editor:
+  - "@fabriziopandini"
+creation-date: 2018-01-28
+last-updated: 2018-01-28
+see-also:
+  - KEP 0004
+
+```
+
+## Table of Contents
+
+  * [kubeadm join --master workflow](#kubeadm-join---master-workflow)
+    * [Metadata](#metadata)
+    * [Table of Contents](#table-of-contents)
+    * [Summary](#summary)
+    * [Motivation](#motivation)
+      * [Goals](#goals)
+      * [Non-goals](#non-goals)
+      * [Challenges and Open Questions](#challenges-and-open-questions)
+    * [Proposal](#proposal)
+      * [User Stories](#user-stories)
+        * [Add a new master node](#add-a-new-master-node)
+      * [Implementation Details](#implementation-details)
+        * [advertise-address = IP/DNS of the external load balancer](#advertise-address--ipdns-of-the-external-load-balancer)
+        * [kubeadm init --feature-gates=HighAvailability=true](#kubeadm-init---feature-gateshighavailabilitytrue)
+        * [kubeadm join --master workflow](#kubeadm-join---master-workflow-1)
+        * [Strategies for deploying control plane components](#strategies-for-deploying-control-plane-components)
+        * [Strategies for distributing cluster certificates](#strategies-for-distributing-cluster-certificates)
+    * [Graduation Criteria](#graduation-criteria)
+    * [Implementation History](#implementation-history)
+    * [Drawbacks](#drawbacks)
+    * [Alternatives](#alternatives)
+
+## Summary
+
+We are extending the kubeadm distinctive `init` and `join` workflow, introducing the
+capability to add more than one master node to an existing cluster by means of the
+new `kubeadm join --master` option.
```

Comment:
I understand that for historic reasons this is what has been decided, but I believe that for the purpose of the KEP, it's worth discussing an alternative option where:
a) TLS assets (and potentially all config files, like static pods) are pre-provisioned, so that the bootstrap process is very deterministic
b) all master nodes run exactly the same command

Which I would see being done like this:
a) a set of TLS assets is created and packaged in a tarball or encoded otherwise
  - user would have to have a list of static IPs and/or DNS names for master nodes
  - the assets would be uploaded to some secure storage provider or copied onto each of the machines
b) on each master user runs `kubeadm init [flags] <assetsConfigPathOrURL>`

Reply:
I guess what I'm suggesting is a step towards an automated implementation of what's [currently documented](kubernetes.io/docs/setup/independent/high-availability/), that doesn't involve manual creation of certificates and could be based around `MasterConfiguration` as input and tarball of `/etc/kubernetes` as output. I believe this would be a simpler option to start with and far more useful to many folks (including our team).

Reply:
I think that such kind of discussion should be extended to a wider audience in the next kubeadm office hours meeting.

I anticipate some comment here:
- I think that nothing in this proposal prevents from pre-provision TLS assets before; however, for simplicity, I prefer to delegate this task to the initial master node
- I like the idea of kubeadm accessing some paths/URL during the join phase for retrieving certs (as alternative to the copy), even if I expect that some kind of security should be defined around this.
- While most of the documented approach today execute `kubeadm init` on all masters, I strongly believe we should deprecate - and possibly block - this at least for two reasons:
  - The init sequence executes a lot of steps with are un-necessary on a secondary master; now those steps are mostly indepondent, so basically now no warm is done by executing them two or three times (some gotchas already known, but nothing blocking so far), but I think this is not sustainable in the long term
  - the init sequence doesn't go through the TLS boostrap process, and this expose the cluster to serious security concerns. IMO any additional master should be forced to go through the TLS bootstrap process, ideally with a dedicated token (different than the token used for joining nodes)
- I think that embedding in TLS assets static IPs and/or DNS names for master nodes is a strategy that prevents/make really complex any future changes to the master nodes (see alternative approach described in this KEP)